### ISSUE TITLE
kube-mgmt is not deleting unexistent k8s resources from opa cache after restart

### ISSUE THREAD

**Issue by muzcategui1106-gs** [open-policy-agent/kube-mgmt#43]
I am using kube-mgmt as sidecar to OPA to sync some kubernetes resources into OPA. I am currently facing an issue where some ingresses that no longer exist in the cluster still show up in OPAs data cache. kube-mgmt continously crashes +30 times in 2 days due to issues with the informer that it uses (see stack trace at the end). I am unable to replicate since I cannot make kube-mgmt purposely crash, but the observed timeline of the issue seems to be as follows

•	An ingress is created   
•	kube-mgmt syncs the ingress
•	Ingress shows up in OPA data cache
•	After X amount of time, kube-mgmt informer gets stuck
•	Ingress gets deleted
•	kube-mgmt panics (see stack trace)
•	kube-mgmt comes back up
•	kube-mgmt syncs resources, but it doesnt delete the deleted ingress from opa's cache

This makes sense since the ingress doesnt exist in the cluster after kube-mgmt comes back up, so the informer never gets a notice to delete the ingress.

Possible solutions

1.	Add ability for kube-mgmt to  terminate the OPA container in the same pod if it crashes for any reason. This will assure that OPA's cache is completely in sync with kube-mgmt
2.	Update the client version that kube-mgmt uses. It currently supports 2.0. The latest client-go version is 11.0. It is recommended that share informers are used to reduce load on the API server

Details:
Kubernetes version: v1.12.9
OPA version: 0.11.0
Kube-mgmt version: 0.8

```
kube-mgmt stacktrace

      Message:   nt-go/tools/cache.NewInformer.func1(0x1349e40, 0xc420fdf3c0, 0x1349e40, 0xc420fdf3c0)
                 /go/src/github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache/controller.go:317 +0x516
github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc42039dd40, 0xc4203edfb0, 0x0, 0x0, 0x0, 0x0)
  /go/src/github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:451 +0x27e
github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache.(*controller).processLoop(0xc42031d200)
  /go/src/github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache/controller.go:147 +0x40
github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache.(*controller).(github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache.processLoop)-fm()
  /go/src/github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache/controller.go:121 +0x2a
github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420568fb0)
  /go/src/github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0x5e
github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4205a3fb0, 0x3b9aca00, 0x0, 0x12a6701, 0xc4201ad8c0)
  /go/src/github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98 +0xbd
github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420568fb0, 0x3b9aca00, 0xc4201ad8c0)
  /go/src/github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x4d
github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache.(*controller).Run(0xc42031d200, 0xc4201ad8c0)
  /go/src/github.com/open-policy-agent/kube-mgmt/vendor/k8s.io/client-go/tools/cache/controller.go:121 +0x237
created by github.com/open-policy-agent/kube-mgmt/pkg/data.(*GenericSync).Run
  /go/src/github.com/open-policy-agent/kube-mgmt/pkg/data/generic.go:80 +0x3b6
```

**Comment by tsandall** [comment]
A couple thoughts.

* We should probably update the client-go dependency since it's been quite a while. I don't think the currently vendored package is still supported. This might help with the problem but without digging into the code, this is just wishful thinking.

* Another option would be to have kube-mgmt wipe out the /kubernetes data in OPA when a resync occurs. This way we'd be guaranteed that stale resources get removed. I'm not sure that client-go indicates when a resync is happening so we may have to implement this ourselves ontop of the lister.

**Comment by muzcategui1106-gs** [comment]
It is actually very easy to replicate this issue, I dont know why I did not think about it before. To simulate simply separate kube-mgmt into a different pod

1. Create a opa pod and service
2. Create kube-mgmt pod. It should point to the opa service and sync ingresses
3. Create Ingress1, it should sync into the opa pod
4) Scale kube-mgmt to 0 replicas
5) Delete ingress1. At this point ingress1 stays in OPA
6) Scale kube-mgmt back to 1 replica
7) Create ingress2. 

At this point Ingress2 will be synced into OPA. You will also see Ingress1 in OPA data cache, eventho it doesnt exist in the cluster anymore. While this scenario is not an exact copy of my deployment (i have kube-mgmt and opa in the same pod), it does simulate what happens if kube-mgmt  contianer continously crashes and the opa container does not.

**Comment by muzcategui1106-gs** [comment]
@tsandall 

I agree, those libraries should be updated regardless. 

I am not sure if kube-mgmt should wipe out the entire data document for OPA as there  could be other systems that put stuff in it. One step forward is to make the informer wipe only the part of the data document managed by itself and resync.

Thoughts?

**Comment by tsandall** [comment]
Wiping out just replicated path(s) is what I had in mind (but my message was a bit unclear.) If people are loading data into replicated path(s) (e.g., `kubernetes/ingresses`)...too bad.

**Comment by tsandall** [comment]
@muzcategui1106-gs I've opened a PR that updates kube-mgmt to use client-go 0.11. Can you give this a try and let me know if it resolves the panic issue you were seeing? https://github.com/open-policy-agent/kube-mgmt/pull/44. Assuming the panics are resolved the original issue of OPA retaining stale resources should be resolved (since kube-mgmt and OPA share fate by running in the same pod, as long as kube-mgmt is not missing updates, OPA will not miss updates.)

**Comment by muzcategui1106-gs** [comment]
@tsandall sorry it has been hectic the last couple of weeks and I hadnt had the chance to chekc on this. I will put it today and leave it running for a couple of days, I can then check if there are no panic issues Does that sound like a sensible way of testing this? I have no way of actually trigger the panic issue on command.

**Comment by tsandall** [comment]
@muzcategui1106-gs yep that sounds fine. I assume you expect resources like ingresses to get deleted like you reported before. Also, you don't have to configure this opa/kube-mgmt deployment to act as the admission controller. As long as kube-mgmt is running and replicating data into opa that should be sufficient (i.e., this shouldn't have to impact the environment.)

**Comment by tsandall** [comment]
There's a tag on docker hub that you can use to test this: `openpolicyagent/kube-mgmt:0.9-rc1`.

**Comment by tsandall** [comment]
This should be fixed by #44
